plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/DevWatchContext.java

```diff
@@ -0,0 +1,522 @@
+            return resolved;
+        }
+
+    }
+
+    private Path getResourcesDir(Path p) {
+        for (Path path : resourceDirectories) {
+            if (p.startsWith(path)) {
+                return path;
+            }
+        }
+        if (p.startsWith(javaDir) && !isJavaFile(p)) {
+            return javaDir;
+        }
+        return null;
+    }
+
+    private boolean isBootableSpecificFile(Path p) {
+        if (cliFiles.contains(p)) {
+            return true;
+        }
+        for (Path path : extraDirectories) {
+            if (p.startsWith(path)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void registerDir(Path dir) throws IOException {
+        registerDir(dir, null);
+    }
+
+    private void registerDir(Path dir, Set<Path> set) throws IOException {
+        Files.walkFileTree(dir, new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                if (!projectBuildDir.equals(dir)) {
+                    watchedDirectories.put(dir.register(watcher, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY), dir);
+                    ctx.debug("[WATCH] watching " + dir);
+                    if (set != null) {
+                        set.add(dir);
+                    }
+                    return FileVisitResult.CONTINUE;
+                } else {
+                    return FileVisitResult.SKIP_SUBTREE;
+                }
+            }
+        });
+    }
+
+    void cleanup() {
+        for (WatchKey k : watchedDirectories.keySet()) {
+            k.cancel();
+        }
+        if (Files.exists(getTargetDirectory())) {
+            IoUtils.recursiveDelete(getTargetDirectory());
+        }
+    }
+
+    BootableAppEventHandler newEventHandler() {
+        BootableAppEventHandler handler;
+        if (isWebApp) {
+            handler = new WebAppEventHandler();
+        } else {
+            if (isJarApp) {
+                handler = new JavaAppEventHandler();
+            } else {
+                throw new RuntimeException("Not supported packaging");
+            }
+        }
+        return handler;
+    }
+
+    void build(boolean autoCompile) throws IOException, MojoExecutionException {
+        rebuild(autoCompile, true, true, true, true);
+    }
+
+    private void rebuild(boolean autoCompile, boolean compile, boolean repackage, boolean redeploy, boolean cleanup) throws IOException, MojoExecutionException {
+        if (cleanup) {
+            ctx.cleanup();
+        }
+        if (compile || cleanup) {
+            ctx.compile(autoCompile);
+        }
+        if (repackage || cleanup) {
+
+            ctx.debug("[WATCH] re-package");
+            if (!Files.exists(ctx.getDeploymentsDir())) {
+                Files.createDirectories(ctx.getDeploymentsDir());
+            }
+            if (isWebApp) {
+                ctx.packageWar(targetDir);
```

Comment:
Just a generic comment that is more of an idea than anything really needing to be changed. Should there just be a `DevWatchContext.packageArchive()` method we could maybe just use a `Consumer<Path>` or `BiConsumer<Path, Path>` where the consumer packages it appropriately based on whether or not it's a web app or JAR? Again just a random thought so no real need to change anything :)